aggregator: introduces retry mechanism

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial committed:
aggregator:introduces retry mechanism

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:
```diff
@@ -107,6 +107,37 @@ func proxyError(w http.ResponseWriter, req *http.Request, error string, code int
+	retryManager := newHijackProtector(w.(*statusResponseWriter), newMaxRetries(newRetryDetector(errRsp), 3))
+
+	for {
+		// TODO: do we have to clone the req ?
```
am pretty sure we have to.

Also, are we sure `w` was not written to?

p0lyn0mial replied:
> Also, are we sure w was not written to?

yes as long as `statusCode and wasHijacked` weren't set 

https://github.com/p0lyn0mial/kubernetes/blob/80f8bf511ceeef0dac6293058c0f80023e15710a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go#L133

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func (hr *hijackResponder) canRetry(err error) bool {
+	if isHTTPVerbRetriable(hr.req) && (knet.IsConnectionReset(err) || knet.IsConnectionRefused(err)) {
```
`IsConnectionReset` is EOF?

p0lyn0mial committed:
ResolveEndpoint: considers already visited EPs

Date:      Mon May 8 12:21:14 2020 +0200

p0lyn0mial committed:
wires ServiceResolverWrapper and ServiceReporter

allows encapsulating possible implementations:
- the old implementation
- a new implementation that only supports retries (ServiceResolverWithVisited)
- a new implementation that supports retries, health reporting and picking up the best possible EP at the given time by examining weights assigned to EPs (ServiceResolverWithVisited, ServiceResolverWithCollector, FailureDetector)

Date:      Tue May 08 12:31:08 2020 +0200

p0lyn0mial committed:
WIP: wire FailureDetector

p0lyn0mial committed:
implements ResolveEndpointWithVisited

p0lyn0mial committed:
fills in serviceReporter

p0lyn0mial committed:
bump(*)

p0lyn0mial committed:
Proxy.Transport.RoundTrip stops suppressing errors
